test/object-store/sync/app.cpp

```diff
@@ -1895,6 +1895,71 @@ TEST_CASE("app: sync integration", "[sync][app]") {
         }
         std::function<util::Optional<Response>(const Request)> hook;
     };
+    SECTION("Fast clock on client") {
+        {
+            TestSyncManager sync_manager(app_config, {});
+            auto app = sync_manager.app();
+            create_user_and_log_in(app);
+            std::shared_ptr<SyncUser> user = app->current_user();
+            SyncTestFile config(app, partition, schema);
+            auto r = Realm::get_shared_realm(config);
+
+            REQUIRE(get_dogs(r).size() == 0);
+            create_one_dog(r);
+            REQUIRE(get_dogs(r).size() == 1);
+        }
+
+        auto transport = std::make_shared<HookedTransport>();
+        app_config.transport = transport;
+
+        TestSyncManager sync_manager(app_config, {});
+        auto app = sync_manager.app();
+        auto creds = create_user_and_log_in(app);
+        std::shared_ptr<SyncUser> user = app->current_user();
+        REQUIRE(user);
+        REQUIRE(!user->access_token_refresh_required());
+        // Make the SyncUser behave as if the client clock is 31 minutes fast, so the token looks expired locallaly
+        // (access tokens have an lifetime of 30 mintutes today).
+        user->set_seconds_to_adjust_time_for_testing(31 * 60);
+        REQUIRE(user->access_token_refresh_required());
+
+        // This assumes that we make an http request for the new token while
+        // already in the WaitingForAccessToken state.
+        std::vector<SyncSession::State> seen_states;
+        transport->hook = [&](const Request) -> util::Optional<Response> {
+            auto user = app->current_user();
+            REQUIRE(user);
+            for (auto session : user->all_sessions()) {
+                seen_states.push_back(session->state());
+
+                // Prior to the fix for #4941, this callback would be called from an infinite loop, always in the
+                // WaitingForAccessToken state.
+                switch (session->state()) {
+                    case SyncSession::State::Active:
+                        puts("state: Active");
```

Comment:
So do the object-store tests just print stuff to stdout now?

Reply:
There is already a lot of logging around these tests, although it might be going to stderr. If you'd like, I'd be happy to remove this. I just found it helpful for observing the infininte loop when I reintroduced the bug that this is supposed to test, to verify that the test would catch it.

Reply:
Instead, how about checking that there is exactly one WaitingForAccessToken state in seen_states?

Reply:
That is fine for the case where it works, but it just hangs silently without the fix. If we think that  is better than printing (and it seems like that is the view of the room), I'm fine removing it.

Reply:
Hmm, maybe we could fail the test when adding a WaitingForAccessToken state to seen_states if one exists there already?

Reply:
Ok I did something similar, but got rid of the vector and just have a bool now. I made the same change to the existing test below to keep them consistent.